### PR TITLE
[proxy] Follow-up refactoring

### PR DESCRIPTION
**Pull request by funbringer**

### CODE REVIEW

File: proxy/src/auth/backend/console.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+/// Common logic for user handling in API V2.
+/// We reuse this for a mock API implementation in [`super::postgres`].
+pub(super) async fn handle_user<'a, Endpoint, GetAuthInfo, WakeCompute>(
+    client: &mut PqStream<impl AsyncRead + AsyncWrite + Unpin>,
+    endpoint: &'a Endpoint,
+    get_auth_info: impl FnOnce(&'a Endpoint) -> GetAuthInfo,
+    wake_compute: impl FnOnce(&'a Endpoint) -> WakeCompute,
+) -> auth::Result<compute::NodeInfo>
```
Comment (LizardWizzard) [inline proxy/src/auth/backend/console.rs:159]: Though I like the unification, this signature is a bit hard to parse visually. Probably nothing we can do other than wait for async traits

File: proxy/src/auth/backend/postgres.rs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+            .query(query, &[&self.creds.user])
+            .await
+            .map_err(io_error)?;
+
+        match &rows[..] {
+            // We can't get a secret if there's no such user.
+            [] => Err(io_error(format!("unknown user '{}'", self.creds.user)).into()),
+
+            // We shouldn't get more than one row anyway.
+            [row, ..] => {
```
Comment (LizardWizzard) [inline proxy/src/auth/backend/postgres.rs:58]: Should we assert that we got only one row? Or thats impossible?

File: proxy/src/sasl.rs [hunk 3]
```diff
@@ -36,6 +37,20 @@ pub enum Error {
...
+impl UserFacingError for Error {
+    fn to_string_client(&self) -> String {
+        use Error::*;
+        match self {
+            // This constructor contains the reason why auth has failed
+            AuthenticationFailed(s) => s.to_string(),
+            // TODO: add support for channel binding
+            ChannelBindingFailed(_) => "channel binding is not supported".to_string(),
+            ChannelBindingBadMethod(m) => format!("unsupported channel binding method {m}"),
+            _ => "authentication failed".to_string(),
```
Comment (LizardWizzard) [inline proxy/src/sasl.rs:49]: BadCliendMessage and Io error will be mapped to "authentication failed" given that we already have dedicated AuthenticationFailed. Looks a bit inconsistent.

**Review by LizardWizzard — COMMENTED** [review]
Looks good to me. Left a few minor comments

File: proxy/src/auth/backend/postgres.rs (hunk 2 above)
Comment (funbringer) [reply proxy/src/auth/backend/postgres.rs:58]: This is impossible, because we can't possibly have duplicate roles.

File: proxy/src/auth/backend/console.rs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        let mut url = self.endpoint.clone();
+        url.path_segments_mut().push("proxy_get_role_secret");
+        url.query_pairs_mut()
+            .append_pair("project", self.project)
+            .append_pair("role", &self.creds.user);
+
+        // TODO: use a proper logger
+        println!("cplane request: {url}");
+
+        let resp = reqwest::get(url.into_inner()).await.map_err(io_error)?;
```
[the remark this answers is not in the document]
Comment (funbringer) [reply proxy/src/auth/backend/console.rs:109]: Reusing a client is not a refactoring, it's a new feature. I'm not doing this right now.

**Review by LizardWizzard — APPROVED** [review]
Thanks!
